cjdns-admin/src/config.rs

```diff
@@ -0,0 +1,176 @@
+const DEFAULT_PORT: u16 = 11234;
+const DEFAULT_PASSWORD: &'static str = "NONE";
+const DEFAULT_CONFIG_FILE_NAME: &'static str = ".cjdnsadmin";
+
+/// Connection options. Can be loaded from a config file.
+#[derive(Clone, Default, PartialEq, Eq, Debug, Deserialize)]
+pub struct Opts {
+    /// Remote IP address (either IPv4 or IPv6).
+    #[serde(rename = "addr")]
+    pub addr: Option<String>,
+
+    /// Remote UDP port.
+    #[serde(rename = "port")]
+    pub port: Option<u16>,
+
+    /// Password for authentication. If `None`, default "NONE" password is used.
+    #[serde(rename = "password")]
+    pub password: Option<String>,
+
+    /// Optional path to config file (`~/.cjdnsadmin` used by default).
+    #[serde(rename = "cjdnsadminPath")]
+    pub config_file_path: Option<String>,
+
+    /// Anonymous connection - do not use password.
+    #[serde(rename = "anon", default)]
+    pub anon: bool,
+}
+
+impl Opts {
+    pub(super) async fn into_connection_options(self) -> Result<ConnectionOptions, Error> {
+        // Do we need to try to read config file?
+        let is_configured = (self.addr.is_some() || self.port.is_some() || self.password.is_some()) && self.config_file_path.is_none();
+
+        // Options to use
+        let mut opts = self;
+        let mut conf_file = None;
+
+        // Try to read config file
+        if !is_configured {
+            if let Some(config_file) = opts.get_config_file_location() {
+                if let Some(config) = Self::read_optional_config_file(&config_file).await? {
+                    opts = config;
+                    conf_file = Some(config_file);
+                }
+            }
+        }
+
+        // Build resulting options
+        Ok(Self::build_connection_options(opts, conf_file))
+    }
+
+    fn build_connection_options(self, conf_file: Option<PathBuf>) -> ConnectionOptions {
+        ConnectionOptions {
+            addr: self.addr.as_ref().map_or(DEFAULT_ADDR, |s| &s).to_string(),
+            port: self.port.unwrap_or(DEFAULT_PORT),
+            password: self.password.as_ref().map_or_else(|| if self.anon { "" } else { DEFAULT_PASSWORD }, |s| &s).to_string(),
+            used_config_file: conf_file.map(|path| path.to_string_lossy().into_owned()),
+        }
+    }
+
+    fn get_config_file_location(&self) -> Option<PathBuf> {
+        if let Some(ref cfg_file) = self.config_file_path {
+            return Some(PathBuf::from(cfg_file));
+        }
+
+        if let Some(mut path) = dirs::home_dir() {
+            path.push(DEFAULT_CONFIG_FILE_NAME);
+            return Some(path.into());
+        }
+
+        None // Unable to locate HOME dir - unsupported platform?
+    }
+
+    fn parse_config(json: &[u8]) -> Result<Self, Error> {
+        serde_json::from_slice(json).map_err(|e| Error::BadConfigFile(e))
+    }
+
+    async fn read_config_file(file_path: &Path) -> Result<Self, Error> {
+        let json = fs::read(file_path).await.map_err(|e| Error::ConfigFileRead(e))?;
```

Comment:
Is it worth to read a local file using async?

Reply:
Why not? If everything else in this crate is async.